### ISSUE TITLE
[Metricbeat Autodiscover] error parsing URL: empty host

### ISSUE THREAD

**Issue by simioa** [elastic/beats#8264]
Metricbeat Version used: 6.4.0, 6.6.2

hints-based autodiscover generates incomplete configurations when there are ports defined in the pod spec and hints include hard-coded ports. Some of them must match. If not, errors related to incorrect hosts happen.

> 2019-03-15T11:28:59.316Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:195	Check failed for config &{{<nil> } <nil> 0xc420681f60}: 3 errors: host parsing failed for etcd-leader: error parsing URL: empty host; host parsing failed for etcd-self: error parsing URL: empty host; host parsing failed for etcd-store: error parsing URL: empty host, won't start runner

`raw` annotations can be used as workaround as mentioned in [this comment](https://github.com/elastic/beats/issues/8264#issuecomment-472669707).

----
Original report:

Hi,

I'm trying to get the Prometheus Collector working with Hints based autodiscovery in Kubernetes.

Unfortunately I always get the following Error:

> ERROR	[autodiscover]	cfgfile/list.go:104	Error creating runner from config: 1 error: 1 error: host parsing failed for prometheus-collector: error parsing URL: empty host

I checked the start event for the host field and it's present.

`Got a start event: map[host:100.123.3.45 port:1234 kubernetes:{"annotations":{"co":{"elastic":{"metrics/hosts":"${data.host}:10254","metrics/metricsets":"collector","metrics/module":"prometheus","metrics/period":"1m"}}}} config:[0xc42001f4a0] start:true], generated configs: [0xc42001f4a0]`

I'm using the following Annotations:
```
      annotations:
        co.elastic.metrics/hosts: ${data.host}:10254
        co.elastic.metrics/metricsets: collector
        co.elastic.metrics/module: prometheus
        co.elastic.metrics/period: 1m
        prometheus.io/port: "10254"
        prometheus.io/scrape: "true"
```

Any idea why this is happening?

**Comment by exekias** [comment]
Hi @simioa , we try to keep Github issues for confirmed bugs and enhancement requests only. Could you please open a new thread in our discuss for this? https://discuss.elastic.co/c/beats

**Comment by MnrGreg** [comment]
Hi @simioa 

I'm getting the same error as you. I noticed your post on discuss.elastic.co as well but didn't see a solution. Did you get this resolved?

I've tried with metricbeat 6.6.1. My hint is the same:

```
      annotations:
        co.elastic.metrics/module: prometheus
        co.elastic.metrics/metricsets: collector
        co.elastic.metrics/hosts: "${data.host}:10254"
        co.elastic.metrics/period: 1m
```

From within the metricbeat container shell I'm able to hit the nginx container prometheus endpoint successfully. Yet still getting the error `host parsing failed for prometheus-collector: error parsing URL: empty host, won't start runner` it seems like metricbeat is not interpreting ${data.host}.


> 2019-03-03T21:15:10.519Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:165	Got a start event: map[meta:{"kubernetes":{"container":{"name":"nginx-ingress-controller"},"labels":{"app":{"kubernetes":{"io/name":"ingress-nginx","io/part-of":"ingress-nginx"}},"pod-template-hash":"1428717591"},"namespace":"kube-ingress-nginx","node":{"name":"k8srck-vpcn303"},"pod":{"name":"nginx-ingress-controller-586dc5c9f5-h6wk2","uid":"6c54ca80-3df9-11e9-99e6-0242ac110003"},"replicaset":{"name":"nginx-ingress-controller-586dc5c9f5"}}} config:[0xc420686480] provider:4818bfc6-be3f-48f3-80c4-f17882ed5f2e id:6c54ca80-3df9-11e9-99e6-0242ac110003.nginx-ingress-controller start:true host:10.244.70.152 port:80 kubernetes:{"annotations":{"co":{"elastic":{"metrics/hosts":"${data.host}:10254","metrics/metrics_path":"/metrics","metrics/metricsets":"collector","metrics/module":"prometheus","metrics/namespace":"kube-ingress-nginx","metrics/period":"1m"}}},"container":{"id":"f4b9a25501277b275ae4c39a0a6c3d3c68143b2b329338296546499410f695a0","image":"container-registry01.wsgc.com/system/nginx-ingress-controller:0.20.0","name":"nginx-ingress-controller","runtime":"docker"},"labels":{"app":{"kubernetes":{"io/name":"ingress-nginx","io/part-of":"ingress-nginx"}},"pod-template-hash":"1428717591"},"namespace":"kube-ingress-nginx","node":{"name":"k8srck-vpcn303"},"pod":{"name":"nginx-ingress-controller-586dc5c9f5-h6wk2","uid":"6c54ca80-3df9-11e9-99e6-0242ac110003"},"replicaset":{"name":"nginx-ingress-controller-586dc5c9f5"}}]
> 2019-03-03T21:15:10.519Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:183	Generated configs: [0xc420686480]
> 2019-03-03T21:15:10.519Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:243	Got a meta field in the event
> 2019-03-03T21:15:10.519Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:195	Check failed for config &{{<nil> } <nil> 0xc42058a140}: 1 error: host parsing failed for prometheus-collector: error parsing URL: empty host, won't start runner
> 2019-03-03T21:15:10.519Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:165	Got a start event: map[id:6c54ca80-3df9-11e9-99e6-0242ac110003.nginx-ingress-controller start:true host:10.244.70.152 port:443 kubernetes:{"annotations":{"co":{"elastic":{"metrics/hosts":"${data.host}:10254","metrics/metrics_path":"/metrics","metrics/metricsets":"collector","metrics/module":"prometheus","metrics/namespace":"kube-ingress-nginx","metrics/period":"1m"}}},"container":{"id":"f4b9a25501277b275ae4c39a0a6c3d3c68143b2b329338296546499410f695a0","image":"container-registry01.wsgc.com/system/nginx-ingress-controller:0.20.0","name":"nginx-ingress-controller","runtime":"docker"},"labels":{"app":{"kubernetes":{"io/name":"ingress-nginx","io/part-of":"ingress-nginx"}},"pod-template-hash":"1428717591"},"namespace":"kube-ingress-nginx","node":{"name":"k8srck-vpcn303"},"pod":{"name":"nginx-ingress-controller-586dc5c9f5-h6wk2","uid":"6c54ca80-3df9-11e9-99e6-0242ac110003"},"replicaset":{"name":"nginx-ingress-controller-586dc5c9f5"}} meta:{"kubernetes":{"container":{"name":"nginx-ingress-controller"},"labels":{"app":{"kubernetes":{"io/name":"ingress-nginx","io/part-of":"ingress-nginx"}},"pod-template-hash":"1428717591"},"namespace":"kube-ingress-nginx","node":{"name":"k8srck-vpcn303"},"pod":{"name":"nginx-ingress-controller-586dc5c9f5-h6wk2","uid":"6c54ca80-3df9-11e9-99e6-0242ac110003"},"replicaset":{"name":"nginx-ingress-controller-586dc5c9f5"}}} config:[0xc420a30750] provider:4818bfc6-be3f-48f3-80c4-f17882ed5f2e]
> 2019-03-03T21:15:10.519Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:183	Generated configs: [0xc420a30750]
> 2019-03-03T21:15:10.519Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:243	Got a meta field in the event
> 2019-03-03T21:15:10.519Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:195	Check failed for config &{{<nil> } <nil> 0xc42058b3e0}: 1 error: host parsing failed for prometheus-collector: error parsing URL: empty host, won't start runner

Any advice?

**Comment by MnrGreg** [comment]
For those interested is seems both hosts and metricsets are lists.

I was able to work around this using the co.elastic.metrics/raw definition:

`co.elastic.metrics/raw: '[{"metricsets":["collector"],"module":"prometheus","namespace":"mynamespace","period":"1m","metrics_path":"/metrics","hosts":["${data.host}:10254"]}]'`

**Comment by jsoriano** [comment]
I can reproduce this issue with metricsets using `URLHostParserBuilder`, I am reopening this and updating description.

**Comment by jsoriano** [comment]
The issue is not in `URLHostParserBuilder` as I previously thought. With other modules this can be also an issue but the reported errors are different and can appear in different moments.
I have found the cause in metricbeat generating configurations with empty hosts when hints-based autodiscover is used. If the pod spec has exposed ports, and the hosts hint has hard-coded ports, they have to match, if they don't, they generate these buggy configurations.
I have opened #12086 to avoid creating these incomplete configurations.
